_drafts/building-a-cli-for-firmware-projects.md

```diff
@@ -0,0 +1,515 @@
+---
+title: Building a CLI for Firmware Projects using Invoke
+description:
+  "A walkthrough on how to build a command line interface (CLI) for a firmware
+  project using the Python Invoke package."
+author: tyler
+tags: [python]
+---
+
+<!-- excerpt start -->
+
+Building a small (or large) command line interface (CLI) for a project is a
+great way to get an entire team to build, test, debug, and work with a project
+in the same way using the same set of tools. This post goes into detail about
+how to think about a project's CLI and implementing one using the
+[Invoke](http://www.pyinvoke.org/) Python package.
+
+<!-- excerpt end -->
+
+## Why Build A Project CLI?
+
+By building a CLI for a project, it becomes a single way for any developer,
+script, or continuous integration system to perform the most common operations
+are performed everyday.
+
+For example, it helps convert launching JLinkGDBServer with the appropriate
+settings from a Wiki copy-paste of:
+
+```
+$ JLinkGDBServer -device Cortex-M4 -if SWD -speed 4000 -RTOSPlugin_FreeRTOS
+```
+
+to
+
+```
+invoke jlink
+```
+
+and flashing using GDB from:
+
+```
+$ cd products/blinky/gcc
+$ arm-none-eabi-gdb-py -x gdbinit.jlink ../build/product.elf -x ../../common_gdb_scripts.gdbinit
+```
+
+to
+
+```
+invoke flash
+```
+
+A web service's REST API needs to be stable, easy to use, and self documenting.
+Your project's CLI should meet the same requirements.
+
+## When to Build Project CLI
+
+If you or your teammates have experienced any of the following more than once,
+it may be time to write a CLI for your project.
+
+- You **copy/paste** commands from a Wiki page or an old Slack message
+- You found the command you were looking for but it's now **out-of-date**
+- It has become common for teammates to ask in Slack **"What is that
+  command..."**
+- Scripts have **undocumented** arguments and environment variable overrides
+- Two or more teammates have **written similar scripts** for themselves but a
+  common shared version doesn't exist, or is broken.
+- You check out a 3 month-old revision of your project and there is no longer
+  documentation about how to run (factory builds anyone?).
+
+I could go on and on, but I hope the idea is clear. Building a CLI using Invoke
+helps with all of the above issues and more.
+
+## Getting started with Invoke
+
+Invoke is a "task execution tool & library" that works with Python 2.7 and 3.4+,
+and be installed using `pip`. We'll be using it today to build our project's CLI
+tool.
+
+We'll be using a virtual environment, and I highly recommend them.
```

Comment:
I highly recommend using a virtual environment, which is a way to sandbox your python environment for your application